pvlib/iotools/sodapro.py

```diff
@@ -0,0 +1,368 @@
+        or as average irradiance values (W/m^2) (pvlib preferred units)
+    label: {'right', 'left'}, default: None
+        Which bin edge label to label time-step with. The default is 'left' for
+        all time steps except for '1M' which has a default of 'right'.
+    map_variables: bool, default: True
+        When true, renames columns of the Dataframe to pvlib variable names
+        where applicable. See variable CAMS_VARIABLE_MAP.
+
+    Returns
+    -------
+    data: pandas.DataFrame
+        Timeseries data from CAMS Radiation or McClear
+    meta: dict
+        Metadata avaiable in the file.
+
+    See Also
+    --------
+    pvlib.iotools.read_cams, pvlib.iotools.get_cams
+
+    References
+    ----------
+    .. [1] `CAMS Radiation Service Info
+       <http://www.soda-pro.com/web-services/radiation/cams-radiation-service/info>`_
+    .. [2] `CAMS McClear Service Info
+       <http://www.soda-pro.com/web-services/radiation/cams-mcclear/info>`_
+    """
+    meta = {}
+    # Initial lines starting with # contain meta-data
+    while True:
+        line = fbuf.readline().rstrip('\n')
+        if line.startswith('# Observation period'):
+            # The last line of the meta-data section contains the column names
+            names = line.lstrip('# ').split(';')
+            break  # End of meta-data section has been reached
+        elif ': ' in line:
+            meta[line.split(': ')[0].lstrip('# ')] = line.split(': ')[1]
+
+    # Convert latitude, longitude, and altitude values from strings to floats
+    for k_old in list(meta.keys()):
+        k_new = k_old.lstrip().split(' ')[0].lower()
+        if k_new in ['latitude', 'longitude', 'altitude']:
+            meta[k_new] = float(meta.pop(k_old))
+
+    meta['radiation_unit'] = {True: 'Wh/m^2', False: 'W/m^2'}[integrated]
+
+    # Determine the time_step from the meta-data dictionary
+    time_step = SUMMATION_PERIOD_TO_TIME_STEP[
+        meta['Summarization (integration) period']]
+    meta['time_step'] = time_step
+
+    data = pd.read_csv(fbuf, sep=';', comment='#', header=None, names=names)
+
+    obs_period = data['Observation period'].str.split('/')
+
+    # Set index as the start observation time (left) and localize to UTC
+    if (label == 'left') | ((label is None) & (time_step != '1M')):
+        data.index = pd.to_datetime(obs_period.str[0], utc=True)
+    # Set index as the stop observation time (right) and localize to UTC
+    # default label for monthly data is 'right' following Pandas' convention
+    elif (label == 'right') | ((label is None) & (time_step == '1M')):
+        data.index = pd.to_datetime(obs_period.str[1], utc=True)
+
+    # Change index for time_step '1d' and '1M' to be date and not datetime
+    if (time_step == '1d') | (time_step == '1M'):
+        data.index = pd.DatetimeIndex(data.index.date)
+    # For monthly data with 'right' label, the index should be the last
+    # date of the month and not the first date of the following month
```

Comment:
Either way seems reasonable to me (and I personally prefer left-labeling to avoid questions like this), but I wonder what other people think.  As a point of comparison, TMY3 data are right-labeled but without a sub-unit shift like this.  

I also wonder if all this labeling code is worth it -- maybe we should just return the start observation time and be done with it.

Reply:
> maybe we should just return the start observation time and be done with it.

+1

Reply:
I recognize your concern and would perhaps agree if it the service didn't return monthly data.

My main argument is that Pandas labels minute and daily data by the left interval, however, monthly data is labelled right (i.e., last day of the month). The majority of users of monthly data would be required to manually change the index labels, before being able to merge with any other resample data in pandas.Personally I think it adds a lot of functionality that makes the function a lot easier to use

Reply:
FWIW here are some one-liners for converting month-begin data to month-end:
```python
In [20]: s = pd.Series([1,2,3], pd.date_range('2019-01-01', periods=3, freq='MS'))

In [21]: s.index
Out[21]: DatetimeIndex(['2019-01-01', '2019-02-01', '2019-03-01'], dtype='datetime64[ns]', freq='MS')

In [22]: s.resample('M').mean().index
Out[22]: DatetimeIndex(['2019-01-31', '2019-02-28', '2019-03-31'], dtype='datetime64[ns]', freq='M')

# or
In [23]: s.to_period('M').to_timestamp('M').index
Out[23]: DatetimeIndex(['2019-01-31', '2019-02-28', '2019-03-31'], dtype='datetime64[ns]', freq='M')
```

How about removing the `label` parameter but adding a docstring example with (a better version of) the above snippet for the benefit of whatever subset of pvlib users want monthly insolation?  E.g. [this](https://github.com/pvlib/pvlib-python/blob/master/pvlib/modelchain.py#L624), which renders as [this](https://pvlib-python.readthedocs.io/en/stable/generated/pvlib.modelchain.ModelChain.with_sapm.html).